### PR TITLE
Clone fix for arm64 android

### PR DESCRIPTION
**Pull request by alvarofe**
This still does not work but the idea is as follows, let me know how it looks like to you or if you think otherwise.

`gum_stalker_iterator_keep` it emits new code when SVC calls are found to check for clone calls. When this happen, `ctx->thread_clone` is set with `GUM_NEW_THREAD_WHILE_STALKING`. Likewise, after encountering a SVC call afterwards is inserted more code to check if the SVC call was clone to get the new thread id of the child - `gum_stalker_thread_creation` is where child and parent is distinguished. Is here where I would like to return to normal life for the child and within the parent notify with a new event so the user can decide whether or not wants to stalk the new thread. However, I am failing to do so - I dunno how to 'disinfect' the child and leave untouched the parent. 

To work around that, I added a few lines within `gum_exec_ctx_replace_current_block_with` to detect when we are running on the child to return to `start_address` but does not seem to work either.

The normal disinfect routine and unfollow would not work as long as the child and parent share the ctx and I thought setting cpu_context->pc to the real address would be enough but either that is not enough or I don't know how to do it correctly.

### CODE REVIEW

File: tests/run-android-arm64.sh [hunk 1]
```diff
@@ -1,14 +1,14 @@
 #!/bin/sh
 
 arch=arm64
+device=$1
 
 remote_prefix=/data/local/tmp/frida-tests-$arch
 
 gum_tests=$(dirname "$0")
 cd "$gum_tests/../../build/tmp-android-$arch/frida-gum" || exit 1
-. ../../frida-meson-env-macos-x86_64.rc
```
[the remark this answers is not in the document]
Comment (oleavr) [reply tests/run-android-arm64.sh]: It's hardcoded to assume macOS build machine; change it to `linux` on yours. (Would be good to fix this though.)

**Comment by alvarofe** [comment]
btw, does it make sense to add another field to ThreadDetails to expose the name of the thread? Might be useful when wanting to only stalk specific threads by their names, leaving alone those for the binder, etc.

```
diff --git a/bindings/gumjs/gumdukprocess.c b/bindings/gumjs/gumdukprocess.c
index 9700c88e..52d7c1b7 100644
--- a/bindings/gumjs/gumdukprocess.c
+++ b/bindings/gumjs/gumdukprocess.c
@@ -214,6 +214,13 @@ gum_emit_thread (const GumThreadDetails * details,
       GUM_CPU_CONTEXT_READONLY, scope->core);
   duk_put_prop_string (ctx, -2, "context");
 
+  if (details->name)
+    duk_push_string (ctx, (details->name));
+  else
+    duk_push_string (ctx, "null");
+
+  duk_put_prop_string (ctx, -2, "name");
+
   if (_gum_duk_scope_call_sync (scope, 1))
   {
     if (duk_is_string (ctx, -1))
diff --git a/bindings/gumjs/gumv8process.cpp b/bindings/gumjs/gumv8process.cpp
index bd92f871..dfc27ffe 100644
--- a/bindings/gumjs/gumv8process.cpp
+++ b/bindings/gumjs/gumv8process.cpp
@@ -183,6 +183,17 @@ gum_emit_thread (const GumThreadDetails * details,
       _gum_v8_cpu_context_new_immutable (&details->cpu_context, core);
   _gum_v8_object_set (thread, "context", cpu_context, core);
 
+  if (details->name)
+  {
+    _gum_v8_object_set (thread, "name", _gum_v8_string_new_ascii (isolate,
+          details->name), core);
+  }
+  else
+  {
+    _gum_v8_object_set (thread, "name", _gum_v8_string_new_ascii (isolate,
+          "null"), core);
+  }
+
   auto proceed = mc->OnMatch (thread);
 
   _gum_v8_cpu_context_free_later (
diff --git a/gum/backend-linux/gumprocess-linux.c b/gum/backend-linux/gumprocess-linux.c
index 7a31a109..9d2d930c 100644
--- a/gum/backend-linux/gumprocess-linux.c
+++ b/gum/backend-linux/gumprocess-linux.c
@@ -577,6 +577,7 @@ _gum_process_enumerate_threads (GumFoundThreadFunc func,
   GDir * dir;
   const gchar * name;
   gboolean carry_on = TRUE;
+  gchar * thread_name_path;
 
   dir = g_dir_open ("/proc/self/task", 0, NULL);
   g_assert (dir != NULL);
@@ -588,11 +589,18 @@ _gum_process_enumerate_threads (GumFoundThreadFunc func,
     details.id = atoi (name);
     if (gum_thread_read_state (details.id, &details.state))
     {
+      thread_name_path = g_strdup_printf ("/proc/self/task/%lu/comm", details.id);
+      if (!g_file_get_contents (thread_name_path, (gchar **) &details.name, NULL,  NULL))
+        details.name = NULL;
+      else
+        details.name = g_strchomp (details.name);
+
       if (gum_process_modify_thread (details.id, gum_store_cpu_context,
             &details.cpu_context))
       {
         carry_on = func (&details, user_data);
       }
+      g_free (thread_name_path);
     }
   }
 
diff --git a/gum/gumprocess.h b/gum/gumprocess.h
index 8cfa6592..d003949d 100644
--- a/gum/gumprocess.h
+++ b/gum/gumprocess.h
@@ -51,6 +51,7 @@ struct _GumThreadDetails
   GumThreadId id;
   GumThreadState state;
   GumCpuContext cpu_context;
+  gchar *name;
 };
 
 struct _GumModuleDetails

```

File: tests/run-android-arm64.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (alvarofe) [reply tests/run-android-arm64.sh]: sure, will update the pr with a fix for this

**Comment by alvarofe** [comment]
This can be reviewed since I have been testing it against my target along with the below module

```ts
var ThreadsFollowed : {[id: number] : boolean} = {};

function isThreadFollowed(threadId: ThreadId) {
  return ThreadsFollowed[threadId];
}

function FollowThread(threadId:ThreadId, options:StalkerOptions) {
  if (isThreadFollowed(threadId)) {
    return;
  }

  ThreadsFollowed[threadId] = true;
  console.log("[+] Following thread " + threadId);
  Stalker.follow(threadId, options);
}

function UnfollowThread(threadId:ThreadId) {
  if (!isThreadFollowed(threadId)) {
    return;
  }

  delete ThreadsFollowed[threadId];
  console.log("[+] Unfollowing thread " + threadId);
  Stalker.unfollow(threadId);
  Stalker.flush();
}

class StalkerThread {
  constructor(options:StalkerOptions) {
    this.options = options;
  }

  Follow(threadId: ThreadId): void {
    FollowThread(threadId, this.options);
  }
  Unfollow(threadId: ThreadId) : void {
    UnfollowThread(threadId);
  }

  options:StalkerOptions;
}


function PthreadStalker(options:StalkerOptions) : StalkerThread {
  const stalker = new StalkerThread(options);
  const pthreadCreate = Module.getExportByName(null, 'pthread_create');
  Interceptor.attach(pthreadCreate, {
    onEnter(args) {
      //if who calls pthread_create is not being followed skip it
      if (!isThreadFollowed(this.threadId)) {
        return;
      }
      const functionAddress = args[2] as NativePointer;
      Interceptor.attach(functionAddress, {
        onEnter(args) {
          stalker.Follow(this.threadId);
        },
        onLeave(retVal) {
          stalker.Unfollow(this.threadId);
        }
      });
    }
  });
  return stalker;
}

export { PthreadStalker };
```

Which I have been using like this

```ts
const pthreadStalker = PthreadStalker({
  events : {
    compile: true
  },
  onReceive(rawEvents) {
    const events = Stalker.parse(rawEvents, {annotate: false}) as StalkerCompileEventBare[];
    events.forEach(ev => {
      const start = ev[0] as NativePointer;
      if (map.has(start))
        send(start);
    });
  }
});

function InterceptAndStalk(addr: NativePointer) {
  Interceptor.attach(addr, {
    onEnter(args) {
      pthreadStalker.Follow(this.threadId);
    }
  });
}

InterceptAndStalk(base.add(***));
InterceptAndStalk(base.add(***));
InterceptAndStalk(base.add(***));
InterceptAndStalk(base.add(***));
```

File: gum/backend-linux/gumprocess-linux.c [hunk 2]
```diff
@@ -305,6 +305,148 @@ gum_process_get_current_thread_id (void)
   return syscall (__NR_gettid);
 }
 
+gboolean
```
Comment (oleavr) [inline gum/backend-linux/gumprocess-linux.c]: I guess these changes can be reverted now that they're no longer needed?

File: tests/run-android-arm64.sh [hunk 3]
```diff
@@ -1,14 +1,28 @@
...
+  echo "$0 <device>"
+  exit 1
+fi
+
+os=$(uname -s)
+if [ $os = "Linux" ]; then
+  os="linux"
+else
+  os="macos"
+fi
```
Comment (oleavr) [inline tests/run-android-arm64.sh]: Would be good to move this to a script next to it that we source, so we can re-use this from the other scripts too.

(But maybe better to do this in a separate PR?)

File: tests/core/arch-arm64/stalker-arm64.c [hunk 4]
```diff
@@ -1240,6 +1241,33 @@ TESTCASE (follow_thread)
...
+TESTCASE (pthread_create)
+{
+  fixture->sink->mask = (GumEventType) (GUM_COMPILE);
+
+  gum_stalker_follow_me (fixture->stalker, fixture->transformer,
+      GUM_EVENT_SINK (fixture->sink));
+
+  pthread_t thread_;
+  int number = 0;
+  int ret = pthread_create (&thread_, NULL, _inc_integer, (gpointer)&number);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
  int ret = pthread_create (&thread, NULL, increment_integer, (gpointer) &number);
```

File: tests/core/arch-arm64/stalker-arm64.c [hunk 5]
```diff
@@ -1240,6 +1241,33 @@ TESTCASE (follow_thread)
 #endif
 }
 
+static gpointer
+_inc_integer(gpointer number)
+{
+  int* a = (int *)number;
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
  int * number = (int *) data;
```

File: tests/core/arch-arm64/stalker-arm64.c [hunk 6]
```diff
@@ -1240,6 +1241,33 @@ TESTCASE (follow_thread)
 #endif
 }
 
+static gpointer
+_inc_integer(gpointer number)
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
increment_integer (gpointer data)
```

File: tests/core/arch-arm64/stalker-arm64.c [hunk 7]
```diff
@@ -1240,6 +1241,33 @@ TESTCASE (follow_thread)
...
+{
+  int* a = (int *)number;
+  *a += 1;
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
  *number += 1;
```

File: tests/core/arch-arm64/stalker-arm64.c (hunk 6 above)
Comment (oleavr) [reply tests/core/arch-arm64/stalker-arm64.c]: Should move this below the function (newspaper code ordering – read from the top, with details further down).

File: gum/backend-arm64/gumstalker-arm64.c [hunk 8]
```diff
@@ -439,6 +441,64 @@ gum_stalker_finalize (GObject * object)
...
+
+  if (gc->opened_prolog != GUM_PROLOG_NONE)
+    gum_exec_block_close_prolog (block, gc);
+
+  gum_exec_block_open_prolog (block, GUM_PROLOG_FULL, gc);
+
+  gum_arm64_writer_put_call_address_with_arguments (gc->code_writer,
+      GUM_ADDRESS (gum_stalker_clone_checker), 2,
+      GUM_ARG_REGISTER, ARM64_REG_X20,
+      GUM_ARG_ADDRESS, GUM_ADDRESS (block->ctx));
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: How about just saving the flags, doing the compare, and then restoring them in each branch? A full prolog/epilog with helper function here seems like a lot of unnecessary overhead.

File: tests/run-android-arm64.sh (hunk 3 above)
Comment (alvarofe) [reply tests/run-android-arm64.sh]: will do in a separete PR

**Comment by alvarofe** [comment]
Updated with all the comments addressed - let me know if I missed something

File: tests/core/arch-arm64/stalker-arm64.c [hunk 9]
```diff
@@ -1384,6 +1405,14 @@ TESTCASE (performance)
   gum_stalker_dump_counters ();
 }
 
+static gpointer
+increment_integer(gpointer data)
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
increment_integer (gpointer data)
```

File: tests/core/arch-arm64/stalker-arm64.c [hunk 10]
```diff
@@ -1384,6 +1405,14 @@ TESTCASE (performance)
...
+{
+  int * number = (int *) data;
+  *number += 1;
+  return NULL;
+}
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]: This function should be moved right after the test.

File: tests/core/arch-arm64/stalker-arm64.c [hunk 11]
```diff
@@ -1240,6 +1242,25 @@ TESTCASE (follow_thread)
...
+TESTCASE (pthread_create)
+{
+  fixture->sink->mask = (GumEventType) (GUM_COMPILE);
+
+  gum_stalker_follow_me (fixture->stalker, fixture->transformer,
+      GUM_EVENT_SINK (fixture->sink));
+
+  pthread_t thread;
+  int number = 0;
+  int ret = pthread_create (&thread, NULL, increment_integer, (gpointer) &number);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]: Variables should be declared at the beginning of the scope, ordered by first use, with one blank line between them and the first statement. (And if one of them is used to store the return value of the function, it should be first.)

File: tests/core/arch-arm64/stalker-arm64.c [hunk 12]
```diff
@@ -67,6 +68,7 @@ static void unfollow_during_transform (GumStalkerIterator * iterator,
 static gboolean store_range_of_test_runner (const GumModuleDetails * details,
     gpointer user_data);
 static void pretend_workload (GumMemoryRange * runner_range);
+static gpointer increment_integer (gpointer data);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]: Should be in chronological order. (Remember to update this after moving the function re. comment below.)

File: gum/backend-arm64/gumstalker-arm64.c [hunk 13]
```diff
@@ -439,6 +441,45 @@ gum_stalker_finalize (GObject * object)
   G_OBJECT_CLASS (gum_stalker_parent_class)->finalize (object);
 }
 
+static GumVirtualizationRequirements
+gum_exec_block_virtualize_linux_sysenter_insn (GumExecBlock * block,
+                                               GumGeneratorContext * gc)
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Function should be moved right after the function using it. (Newspaper ordering.)

File: gum/backend-arm64/gumstalker-arm64.c [hunk 14]
```diff
@@ -439,6 +441,45 @@ gum_stalker_finalize (GObject * object)
...
+{
+  GumArm64Writer *cw = gc->code_writer;
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  GumArm64Writer * cw = gc->code_writer;
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 15]
```diff
@@ -439,6 +441,45 @@ gum_stalker_finalize (GObject * object)
...
+
+  gum_arm64_writer_put_sub_reg_reg_imm (cw, ARM64_REG_X17,
+      ARM64_REG_X8, __NR_clone);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X17, continue_normally);
+
+  gum_arm64_writer_put_instruction (cw, msr_nzcv_x15);
+  gum_arm64_writer_put_bytes (cw, insn->bytes, 4);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X0, end_payload);
+
+  //we are on the child return to the original next instruction
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  /* We are on the child return to the original next instruction. */
```

Single-line comments should always be C-style and follow this style.

File: gum/backend-arm64/gumstalker-arm64.c [hunk 16]
```diff
@@ -439,6 +441,45 @@ gum_stalker_finalize (GObject * object)
...
+  const cs_insn * insn = gc->instruction->ci;
+  const guint32 mrs_x15_nzcv = 0xd53b420f;
+  const guint32 msr_nzcv_x15 = 0xd51b420f;
+
+  gum_arm64_relocator_skip_one (gc->relocator);
+
+  if (gc->opened_prolog != GUM_PROLOG_NONE)
+    gum_exec_block_close_prolog (block, gc);
+
+  gum_arm64_writer_put_instruction (cw, mrs_x15_nzcv);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Won't this clobber `X15`?

File: gum/backend-arm64/gumstalker-arm64.c [hunk 17]
```diff
@@ -439,6 +441,45 @@ gum_stalker_finalize (GObject * object)
...
+
+  gum_arm64_writer_put_sub_reg_reg_imm (cw, ARM64_REG_X17,
+      ARM64_REG_X8, __NR_clone);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X17, continue_normally);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]: Are we allowed to clobber `X17` here?

File: gum/backend-arm64/gumstalker-arm64.c [hunk 18]
```diff
@@ -439,6 +441,45 @@ gum_stalker_finalize (GObject * object)
   G_OBJECT_CLASS (gum_stalker_parent_class)->finalize (object);
 }
 
+static GumVirtualizationRequirements
+gum_exec_block_virtualize_linux_sysenter_insn (GumExecBlock * block,
+                                               GumGeneratorContext * gc)
+{
+  GumArm64Writer *cw = gc->code_writer;
+  gconstpointer continue_normally = cw->code + 1;
+  gconstpointer end_payload = cw->code + 3;
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  gconstpointer end_payload = cw->code + 2;
```

(These just have to be unique, but should be +1 relative to each order for readability.)

File: gum/backend-arm64/gumstalker-arm64.c (hunk 17 above)
Comment (alvarofe) [reply gum/backend-arm64/gumstalker-arm64.c]: X17 is a scratch register and it's normally used for such purposes for the linker - in fact when doing gum_exec_block_write_exec_generated_code it uses the same, X16 and X17 to return to the next instruction. X15 was clobbered and fixed, but when transfering to the original code X17 has to be clobbered.

File: tests/core/arch-arm64/stalker-arm64.c (hunk 11 above)
Comment (alvarofe) [reply tests/core/arch-arm64/stalker-arm64.c]: Done

File: tests/core/arch-arm64/stalker-arm64.c (hunk 12 above)
Comment (alvarofe) [reply tests/core/arch-arm64/stalker-arm64.c]: Done (at least I think so)

File: gum/backend-arm64/gumstalker-arm64.c (hunk 13 above)
Comment (alvarofe) [reply gum/backend-arm64/gumstalker-arm64.c]: Fixed

File: gum/backend-arm64/gumstalker-arm64.c (hunk 14 above)
Comment (alvarofe) [reply gum/backend-arm64/gumstalker-arm64.c]: Done

File: gum/backend-arm64/gumstalker-arm64.c (hunk 15 above)
Comment (alvarofe) [reply gum/backend-arm64/gumstalker-arm64.c]: Done

File: gum/backend-arm64/gumstalker-arm64.c (hunk 16 above)
Comment (alvarofe) [reply gum/backend-arm64/gumstalker-arm64.c]: Yep, updated pushing X15 to the stack

File: gum/backend-arm64/gumstalker-arm64.c (hunk 17 above)
Comment (alvarofe) [reply gum/backend-arm64/gumstalker-arm64.c]: If there is another way to return back to the original code I am all ears

File: gum/backend-arm64/gumstalker-arm64.c (hunk 18 above)
Comment (alvarofe) [reply gum/backend-arm64/gumstalker-arm64.c]: updated

**Comment by alvarofe** [comment]
> Getting there!

This can be reviewed again

File: gum/backend-arm64/gumstalker-arm64.c (hunk 17 above)
Comment (oleavr) [reply gum/backend-arm64/gumstalker-arm64.c]: I know it's a scratch register, but if we clobber it someone might write some inline assembly to detect Stalker this way.

We don't actually clobber it in `write_exec_generated_code()`, we generate code that saves it before we clobber it.

**Comment by oleavr** [comment]
One thing before I do a full review pass again: it would be good to save register(s) well below the stack pointer like we do elsewhere, to avoid clobbering the red-zone. (This is why we save state right below offset `-128` from `SP`.)

**Comment by alvarofe** [comment]
Updated with the red zone issue, let me know if you see something else missing

File: gum/backend-arm64/gumstalker-arm64.c [hunk 19]
```diff
@@ -2616,7 +2620,58 @@ static GumVirtualizationRequirements
...
+#endif
+}
+
+static GumVirtualizationRequirements
+gum_exec_block_virtualize_linux_sysenter (GumExecBlock * block,
+                                          GumGeneratorContext * gc)
+{
+  GumArm64Writer * cw = gc->code_writer;
+  const cs_insn * insn = gc->instruction->ci;
+  gconstpointer continue_normally = cw->code + 1;
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  gconstpointer perform_regular_syscall = cw->code + 1;
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 20]
```diff
@@ -2616,7 +2620,58 @@ static GumVirtualizationRequirements
...
+  gconstpointer end_payload = cw->code + 2;
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  gconstpointer perform_next_instruction = cw->code + 2;
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 21]
```diff
@@ -2616,7 +2620,58 @@ static GumVirtualizationRequirements
...
+      GUM_ADDRESS (gc->instruction->begin + 4));
+  gum_arm64_writer_put_br_reg (cw, ARM64_REG_X17);
+
+  gum_arm64_writer_put_label (cw, continue_normally);
+  gum_arm64_writer_put_instruction (cw, msr_nzcv_x15);
+  gum_arm64_writer_put_ldp_reg_reg_reg_offset (cw, ARM64_REG_X15,
+      ARM64_REG_X17, ARM64_REG_SP, 16 + GUM_RED_ZONE_SIZE,
+      GUM_INDEX_POST_ADJUST);
+  gum_arm64_writer_put_bytes (cw, insn->bytes, 4);
+  gum_arm64_writer_put_label (cw, end_payload);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion

  gum_arm64_writer_put_label (cw, perform_next_instruction);
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 22]
```diff
@@ -2616,7 +2620,58 @@ static GumVirtualizationRequirements
...
+  gum_arm64_writer_put_sub_reg_reg_imm (cw, ARM64_REG_X17,
+      ARM64_REG_X8, __NR_clone);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X17, continue_normally);
+
+  gum_arm64_writer_put_instruction (cw, msr_nzcv_x15);
+  gum_arm64_writer_put_ldp_reg_reg_reg_offset (cw, ARM64_REG_X15,
+      ARM64_REG_X17, ARM64_REG_SP, 16 + GUM_RED_ZONE_SIZE,
+      GUM_INDEX_POST_ADJUST);
+  gum_arm64_writer_put_bytes (cw, insn->bytes, 4);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X0, end_payload);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X0,
      perform_next_instruction);
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 23]
```diff
@@ -2616,7 +2620,58 @@ static GumVirtualizationRequirements
...
+  if (gc->opened_prolog != GUM_PROLOG_NONE)
+    gum_exec_block_close_prolog (block, gc);
+
+  gum_arm64_writer_put_stp_reg_reg_reg_offset (cw, ARM64_REG_X15, ARM64_REG_X17,
+      ARM64_REG_SP, -(16 + GUM_RED_ZONE_SIZE), GUM_INDEX_PRE_ADJUST);
+  gum_arm64_writer_put_instruction (cw, mrs_x15_nzcv);
+
+  gum_arm64_writer_put_sub_reg_reg_imm (cw, ARM64_REG_X17,
+      ARM64_REG_X8, __NR_clone);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X17, continue_normally);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X17,
      perform_regular_syscall);
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 24]
```diff
@@ -2616,7 +2620,58 @@ static GumVirtualizationRequirements
...
+      GUM_INDEX_POST_ADJUST);
+  gum_arm64_writer_put_bytes (cw, insn->bytes, 4);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X0, end_payload);
+
+  /* We are on the child return to the original next instruction */
+  gum_arm64_writer_put_ldr_reg_address (cw, ARM64_REG_X17,
+      GUM_ADDRESS (gc->instruction->begin + 4));
+  gum_arm64_writer_put_br_reg (cw, ARM64_REG_X17);
+
+  gum_arm64_writer_put_label (cw, continue_normally);
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  gum_arm64_writer_put_label (cw, perform_regular_syscall);
```

File: gum/backend-arm64/gumstalker-arm64.c [hunk 25]
```diff
@@ -2616,7 +2620,58 @@ static GumVirtualizationRequirements
...
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X17, continue_normally);
+
+  gum_arm64_writer_put_instruction (cw, msr_nzcv_x15);
+  gum_arm64_writer_put_ldp_reg_reg_reg_offset (cw, ARM64_REG_X15,
+      ARM64_REG_X17, ARM64_REG_SP, 16 + GUM_RED_ZONE_SIZE,
+      GUM_INDEX_POST_ADJUST);
+  gum_arm64_writer_put_bytes (cw, insn->bytes, 4);
+  gum_arm64_writer_put_cbnz_reg_label (cw, ARM64_REG_X0, end_payload);
+
+  /* We are on the child return to the original next instruction */
```
Comment (oleavr) [inline gum/backend-arm64/gumstalker-arm64.c]:
```suggestion
  /*
   * We are on the child return to the original next instruction.
   *
   * TODO: Is there any way we can avoid clobbering X17 here?
   */
```

Here we're clobbering the child's X17. Can't think of a better way, though, so we should probably leave a comment here.

File: tests/core/arch-arm64/stalker-arm64.c [hunk 26]
```diff
@@ -1240,6 +1242,34 @@ TESTCASE (follow_thread)
...
+  int ret;
+  pthread_t thread;
+  int number = 0;
+  fixture->sink->mask = (GumEventType) (GUM_COMPILE);
+
+  gum_stalker_follow_me (fixture->stalker, fixture->transformer,
+      GUM_EVENT_SINK (fixture->sink));
+
+  ret = pthread_create (&thread, NULL, increment_integer, (gpointer) &number);
+  g_assert_cmpuint (ret, ==, 0);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
  g_assert_cmpint (ret, ==, 0);
```

As it's signed, not unsigned.

File: tests/core/arch-arm64/stalker-arm64.c [hunk 27]
```diff
@@ -1240,6 +1242,34 @@ TESTCASE (follow_thread)
...
+
+  ret = pthread_join (thread, NULL);
+  g_assert_cmpuint (ret, ==, 0);
+
+  g_assert_cmpuint (number, ==, 1);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
  g_assert_cmpint (number, ==, 1);
```

File: tests/core/arch-arm64/stalker-arm64.c [hunk 28]
```diff
@@ -1240,6 +1242,34 @@ TESTCASE (follow_thread)
...
+  fixture->sink->mask = (GumEventType) (GUM_COMPILE);
+
+  gum_stalker_follow_me (fixture->stalker, fixture->transformer,
+      GUM_EVENT_SINK (fixture->sink));
+
+  ret = pthread_create (&thread, NULL, increment_integer, (gpointer) &number);
+  g_assert_cmpuint (ret, ==, 0);
+
+  ret = pthread_join (thread, NULL);
+  g_assert_cmpuint (ret, ==, 0);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion
  g_assert_cmpint (ret, ==, 0);
```

File: tests/core/arch-arm64/stalker-arm64.c [hunk 29]
```diff
@@ -1240,6 +1242,34 @@ TESTCASE (follow_thread)
...
+
+  g_assert_cmpuint (number, ==, 1);
+  gum_stalker_unfollow_me (fixture->stalker);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion

  gum_stalker_unfollow_me (fixture->stalker);
```

File: tests/core/arch-arm64/stalker-arm64.c [hunk 30]
```diff
@@ -1240,6 +1242,34 @@ TESTCASE (follow_thread)
 #endif
 }
 
+TESTCASE (pthread_create)
+{
+  int ret;
+  pthread_t thread;
+  int number = 0;
+  fixture->sink->mask = (GumEventType) (GUM_COMPILE);
```
Comment (oleavr) [inline tests/core/arch-arm64/stalker-arm64.c]:
```suggestion

  fixture->sink->mask = (GumEventType) GUM_COMPILE;
```

**Comment by alvarofe** [comment]
Updated again - next time will try to be more careful with the coding style and being more descriptive with variables name.

**Comment by oleavr** [comment]
Oops, only two to go.

File: tests/core/arch-arm64/stalker-arm64.c (hunk 30 above)
Comment (oleavr) [reply tests/core/arch-arm64/stalker-arm64.c]: ^ This one didn't make it last time.

**Comment by alvarofe** [comment]
Updated once again - hopefully this will be the last iteration ;)

**Comment by oleavr** [comment]
Thanks, merged! 🙌
